vllm/model_executor/models/qwen2_5_vl.py

```diff
@@ -768,6 +884,185 @@ def load_weights(self, weights: Iterable[Tuple[str,
+        return (
+            hidden_states,
+            rotary_pos_emb,
+            cu_seqlens,
+            cu_window_seqlens,
+            window_index,
+        )
+
+    def forward(self, x: torch.Tensor, fullattn_mask: Optional[torch.Tensor],
+                rotary_pos_emb: torch.Tensor) -> torch.Tensor:
+        assert_msg = (
+            "Expect inputs to be 112x112 aligned. "
+            "Please align before sending image or use this version "
+            "of transformer that does the resizing/alignment automatically:"
+            "pip install -r requirements-hpu-qwen2_5_vl.txt")
+        assert x.shape[0] % 64 == 0, assert_msg
+        hidden_states = x.unsqueeze(1)
+        for layer_num, blk in enumerate(self.blocks):
+            htcore.mark_step()
+            hidden_states = blk(hidden_states,
+                                cu_seqlens=fullattn_mask if layer_num
+                                in self.fullatt_block_indexes else None,
+                                rotary_pos_emb=rotary_pos_emb)
+        return hidden_states
+
+    def post_attn(self, hidden_states: torch.Tensor,
+                  window_index: torch.Tensor):
+        # adapter
+        hidden_states = self.merger(hidden_states)
+        reverse_indices = torch.argsort(window_index)
+
+        hidden_states = hidden_states[reverse_indices, :]
+        return hidden_states
+
+    def get_image_embeds(
+        self,
+        pixel_values: torch.Tensor,
+        grid_thw: torch.Tensor,
+        vision_buckets,
+    ) -> torch.Tensor:
+
+        assert pixel_values.shape[0] % 64 == 0, (
+            f"We need image h/w to be aligned to 112 for now. "
+            f"Which will make pixel_values be a multiple of "
+            f"(112/14)*(112/14)=64"
+            f"(14 is patch size for ViT). "
+            f"Got pixel_values shape {pixel_values.shape[0]}")
+        offset = 0
+        results = []
+        # process each image one by one
+        for img_idx in range(grid_thw.shape[0]):
+            img_shape = grid_thw[img_idx, :].unsqueeze(0)
+            curr_img_size = img_shape.prod()
+
+            pixel_values_curr_img = pixel_values[offset:offset +
+                                                 curr_img_size, :]
+
+            offset += curr_img_size
+            pixel_values_curr_img_padded, img_shape_padded = \
+                self.pad_multimodal_data(
+                    pixel_values_curr_img,
+                    img_shape,
+                    vision_buckets=vision_buckets
+                )
+
+            pixel_values_curr_img_padded, rot_pos_emb, \
+                cu_seqlens, _, window_index = self.pre_attn(
+                pixel_values_curr_img_padded, img_shape_padded)
+
+            # either a single image,
+            # or a single image and its accompanying pad image,
+            # so only max expansion to 3
+            expanded_cu_seqlens = expand_to_max(cu_seqlens, 3)
```

Comment:
@jiminha rightly points out this can be removed now.

Earlier we were passing it in to the fwd pass and the attn mask was created inside. now we create it in line 1047

Earlier we wanted shape of cu_seqlens to be same regardless of padding or not padding. if no padding its len was 2 (0 and imagesize), and with pad it was len = (0, imglen, padding), so we expanded it to sizze 3 to prevent recompilations because it was entering forward (where we wanna prevent recompilations)

In later optimizations we create the mask fullatt_block_attn_mask externally (which is faster to do it only once and needs less memory). so we can remove this line and use cu_seqlens[-1] in the assert in line 1048